### DISCUSSION TITLE
SQLAlchemy 2.0 reg.mapped_as_dataclass() and dataclasses.field()

### DISCUSSION THREAD

**Discussion by xlorepdarkhelm** [sqlalchemy/sqlalchemy#9575] [Usage Questions]
I am looking for the best approach to be able to provide an equivalent of the dataclasses.field(metadata={}) metadata parameter using the reg.mapped_as_dataclass() decorator and the mapped_column() column definition.

I am building a dataclass-based table model for SQLAlchemy 2.0, and I am using desert (with marshmallow) to convert that data (with validation capabilities) into a dict structure and/or JSON and back. My process paths are:

SQL table -> SQLAlchemy ORM object (dataclass-based) -> desert-created marshmallow Schema -> dict/JSON
dict/JSON -> desert-created marshmallow Schema -> SQLAchemy ORM object (dataclass-based) -> SQL table

The problem is that desert is apparently unable to identify the correct type from Mapped[str] in order to make the Schema object. It recommends setting the metadata to correctly identify the type. However mapped_column() doesn't have this capability, and I can't use dataclasses.field() with reg.mapped_as_dataclass() classes.

What's the best practice to be able to pass that metadata data to the field so that I can have desert be able to identify the field correctly?

**Comment by zzzeek** [comment]
Hi -

I would need more information here on what the problem with "dessert" is, since the "Mapped[]" datatype is not what's passed to the dataclass constructor.  If "dessert" is looking at ``__dataclass_fields__``, the `Mapped[]` type is not there:

```py
from __future__ import annotations

from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import registry

reg = registry()

@reg.mapped_as_dataclass()
class User:
    __tablename__ = "sys_user"

    id: Mapped[int] = mapped_column(primary_key=True)
    username: Mapped[str] = mapped_column()
    email: Mapped[str] = mapped_column()

print(User.__dataclass_fields__)
```

where we can see that plain ForwardRef types are there:

```
{'id': Field(name='id',type=ForwardRef('int'),default=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,default_factory=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD), 'username': Field(name='username',type=ForwardRef('str'),default=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,default_factory=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD), 'email': Field(name='email',type=ForwardRef('str'),default=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,default_factory=<dataclasses._MISSING_TYPE object at 0x7f93d8e14070>,init=True,repr=True,hash=None,compare=True,metadata=mappingproxy({}),kw_only=False,_field_type=_FIELD)}
```

you can manipulate those Field objects if more is needed, or you can provide a "dataclass_callable" to do it at once:

```py
def wrap_dc(cls: Any, **kw: Any) -> Any:
    for key in dir(cls):
        obj = getattr(cls, key)
        if isinstance(obj, dataclasses.Field):
            obj.metadata = {"dessert_type": obj.type}
    return dataclasses.dataclass(cls, **kw)

reg = registry()

@reg.mapped_as_dataclass(dataclass_callable=wrap_dc)
class User:
    __tablename__ = "sys_user"

    id: Mapped[int] = mapped_column(primary_key=True)
    username: Mapped[str] = mapped_column()
    email: Mapped[str] = mapped_column()
```

We can support adding a new field called "dataclass_metadata" to populate Field.metadata directly but we don't have that right now.

overall this seems like a bug in dessert since there should not be any problem with the datatypes, and additionally even if dessert is consuming the ``Mapped[]`` datatype directly, that's a [descriptor type](https://docs.python.org/3/howto/descriptor.html); dessert should know how to handle this by seeing that it has a ``__get__()`` and ``__set__()`` method and inspecting the datatypes around those methods.

**Comment by xlorepdarkhelm** [comment]
Its not dessert, but desert. https://github.com/python-desert/desert

What desert does is construct a marshmallow Schema object based on a dataclass definition. This then allows marshmallow to marshal/unmarshal the object into a dictionary structure or a JSON string, and provides validation of the data during the process. Marshmallow is a fairly common Python library (even if desert isn't necessarily as such right now).

Desert does this to keep marshmallow DRY. Otherwise you have to construct a second class just to have as a Schema to control the conversion. Desert's methodology lets me write 1 class, and it does it all. But the class needs to be a dataclass or attrs. It does it's magic through the type annotations for the class in question to attempt to figure out what data types to use for the Schema object. But, if it cannot understand the type annotations, it has an alternative method for accomplishing this through the field metadata parameter, where it is possible to specify which marshmallow fIeld type to use.

SQLAlchemy's Declarative dataclasses do not have a simple way to provide access to the field's metadata. Desert is not the only library that expects that to be available for dataclasses.

I guess what I am hoping for is what you suggested: Adding a dataclass_metadata field to populate the Field.metadata directly. As this is something that would be likely used by multiple dataclass-based libraries, I can see value in SQLAlchemy having this functionality.

**Reply by zzzeek** [reply]
> Its not dessert, but desert. https://github.com/python-desert/desert
> 
> What desert does is construct a marshmallow Schema object based on a dataclass definition. This then allows marshmallow to marshal/unmarshal the object into a dictionary structure or a JSON string, and provides validation of the data during the process. Marshmallow is a fairly common Python library (even if desert isn't necessarily as such right now).

sure, but as I mentioned, if it sees a dataclasses `Field` object in the first place at runtime (which is when SQLAlchemy makes them), the type that it wants is in that `Field` already.   Field.type is public, it's documented, desert should absolutely be using that, and not even bothering with `__annotations__`:

https://docs.python.org/3/library/dataclasses.html#dataclasses.Field

However in either case, it should also understand what a descriptor type is and how to interpret it, which is what `Mapped[]` is.

These are bugs in desert that should be reported.

> 
> SQLAlchemy's Declarative dataclasses do not have a simple way to provide access to the field's metadata. Desert is not the only library that expects that to be available for dataclasses.

SQLAlchemy's Declarative dataclasses **are actual Python dataclasses**.   Any library that interprets dataclasses will not see any difference with a SQLAlchemy class.   The Field objects on a dataclass are only available at runtime after class creation time from the ``__dataclass_fields__`` attribute.  That is normal:

```py
import dataclasses

@dataclasses.dataclass
class Foo:
    id: int

    bar: str = dataclasses.field(default_factory=lambda: 'hi')

try:
    print(Foo.bar)
except AttributeError as ae:
    print(f"dataclasses do not have class-level attributes: {ae}")

print(f"The dataclasses fields are here, just like they are in SQLAlchemy: {Foo.__dataclass_fields__}")
```

There is **no other way** that desert could be accessing these fields unless you are using a class decorator that is custom to the "desert" library, rather than `@dataclasses.dataclass`

**Comment by xlorepdarkhelm** [comment]
i did find the undocumented fields `__dataclass_fields__` and `__dataclass_params__`. I am concerned that as these are (from what I can tell) undocumented, they might be up for being altered in the future, and I am leery of them. That said, I am attempting to use those to get the effect of what I need through them currently.

**Reply by zzzeek** [reply]
as mentioned above, desert has to be using this attribute because those `Field` objects are nowhere else once the class is created.    

There's the choice of 1. fix bugs in desert, everything works with no more efforts 2. get a new feature added to SQLAlchemy to work around the bugs in desert which would also require that you write a cumbersome and redundant "dataclasses_metadata={}" directive for all of your fields.

however, because I still dont know how "desert" accesses these Field objects, it's not clear to me that a SQLAlchemy feature would even work here.   if "desert" is already accessing the Field objects, it already has the information it needs.

**Reply by CaselIT** [reply]
the fields are also available using https://docs.python.org/3/library/dataclasses.html#dataclasses.fields

**Reply by zzzeek** [reply]
ah thanks didn't know where that was
